File: plugins/plugin-code/typescript/actions/readFile.ts
import * as path from "node:path";
import type {
  Action,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { CoderService } from "../services/coderService";

function extToLang(ext: string): string {
  const e = ext.toLowerCase();
  if (e === ".ts" || e === ".tsx") return "ts";
  if (e === ".js" || e === ".jsx") return "js";
  if (e === ".json") return "json";
  if (e === ".md") return "md";
  if (e === ".py") return "py";
  if (e === ".rs") return "rs";
  if (e === ".go") return "go";
  if (e === ".toml") return "toml";
  if (e === ".yml" || e === ".yaml") return "yaml";
  if (e === ".sh" || e === ".bash" || e === ".zsh") return "bash";
  return "";
}

function getFilepath(
  options: HandlerOptions | undefined,
  message: Memory,
): string {
  const opt = options as { filepath?: string } | undefined;
  if (opt?.filepath && opt.filepath.trim().length > 0)
    return opt.filepath.trim();
  const text = message.content.text ?? "";
  const m = text.match(/["'`]([^"'`]+)["'`]/);
  if (m?.[1]) return m[1].trim();
  const loose = text.match(/(?:\.\/|\/)?[\w\-./]+\.[a-zA-Z0-9]+/);
  return loose?.[0]?.trim() ?? "";
}

const spec = requireActionSpec("READ_FILE");

export const readFile: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: `${spec.description} Reads a file's contents.`,
  validate: async (
    runtime: IAgentRuntime,
    _message: Memory,
  ): Promise<boolean> => runtime.getService<CoderService>("coder") !== null,
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    options: HandlerOptions | undefined,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const svc = runtime.getService<CoderService>("coder");
    if (!svc) {
      const msg = "Coder service is not available.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const filepath = getFilepath(options, message);
    if (!filepath) {
      const msg = "Missing filepath.";
      if (callback) await callback({ content: { text: msg } });
      return { success: false, text: msg };
    }

    const conv = message.roomId ?? message.agentId ?? runtime.agentId;
    const result = await svc.readFile(conv, filepath);
    if (!result.ok) {
      if (callback) await callback({ content: { text: result.error } });
      return { success: false, text: result.error };
    }

    const lang = extToLang(path.extname(filepath));
    const fenced = `\`\`\`${lang}\n${result.content}\n\`\`\``;
    if (callback) await callback({ content: { text: fenced } });
    return { success: true, text: fenced };
  },
};

File: plugins/plugin-code/rust/src/actions/read_file.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::service::CoderService;
use crate::{Action, ActionExample, ActionResult};

pub struct ReadFileAction;

#[async_trait]
impl Action for ReadFileAction {
    fn name(&self) -> &str {
        "READ_FILE"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "VIEW_FILE",
            "OPEN_FILE",
            "CAT_FILE",
            "SHOW_FILE",
            "GET_FILE",
        ]
    }

    fn description(&self) -> &str {
        "Read and return a file's contents."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&mut CoderService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "Coder service is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let filepath = state
            .get("filepath")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        if filepath.trim().is_empty() {
            return ActionResult {
                success: false,
                text: "Missing filepath.".to_string(),
                data: None,
                error: Some("missing_filepath".to_string()),
            };
        }

        let conv = message
            .get("room_id")
            .and_then(|v| v.as_str())
            .or_else(|| message.get("agent_id").and_then(|v| v.as_str()))
            .unwrap_or("default");

        match svc.read_file(conv, &filepath).await {
            Ok(content) => ActionResult {
                success: true,
                text: content,
                data: None,
                error: None,
            },
            Err(err) => ActionResult {
                success: false,
                text: err,
                data: None,
                error: Some("read_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![ActionExample {
            user_message: "read src/lib.rs".to_string(),
            agent_response: "Reading file…".to_string(),
        }]
    }
}
